### PR TITLE
Open Price Feed Poster Clean-up

### PR DESCRIPTION
**Pull request by hayesgm**
This patch is simply cleaning up some of the Open Price Feed code. I started and I couldn't stop, so it's a nice overhaul just trying to cleanup the functions and interface without significantly changing the behavior. There are some minor CLI differences that we'll need to account for. This is going to be the basis for the branch to handle updating Uniswap for testnets.

### CODE REVIEW

File: poster/src/poster.ts [hunk 1]
```diff
@@ -31,195 +42,121 @@ async function main(sources: string,
...
+            },
+            messages: [...acc.messages, payload.messages[index]],
+            signatures: [...acc.signatures, payload.signatures[index]]
+          };
+        } else {
+          return acc;
+        }
+      } else {
+        // Post only prices for supported assets, skip prices for unregistered assets
+        console.debug(`Skipping ${asset} as not part of supported assets: ${JSON.stringify(supportedAssets)}`);
```
Comment (coburncoburn) [inline poster/src/poster.ts:90]: Nit, may be more clear to filter the assets before or after this reduce action like
```
Object.entries(payload.prices).filter( asset => supportedAssets.includes(asset.toUpperCase())).reduce...
```
loses the logging though 🤷

File: poster/src/poster.ts [hunk 2]
```diff
@@ -31,195 +42,121 @@ async function main(sources: string,
...
-        filteredSignatures.push(payload.signatures[index]);
-      }
-      index++;
+  let filteredPayloads = await Promise.all(payloads.map(async payload => {
+    let sourceAddresses = await Promise.all(payload.messages.map((_, i) => {
+      return getSourceAddress(dataAddress, payload.messages[0], payload.signatures[0], web3);
+    }));
+
+    if ([...new Set(sourceAddresses)].length !== 1) {
+      throw new Error(`Invalid source addresses, got: ${JSON.stringify(sourceAddresses)}`);
```
Comment (coburncoburn) [inline poster/src/poster.ts:60]: interesting requirement, fine with me, but why did this come up?

File: poster/src/poster.ts [hunk 3]
```diff
@@ -31,195 +42,121 @@ async function main(sources: string,
...
+  const difference = new BN(prevPrice).minus(new BN(price).multipliedBy(1e6)).abs();
+
   return difference.isLessThanOrEqualTo(minDifference);
 }
 
-async function fetchPayloads(sources: string[], fetchFn = fetch): Promise<DelFiReporterPayload[]> {
-  let sourcePromises = sources.map(async (source) => {
+export async function fetchPayloads(sources: string[], fetchFn=fetch): Promise<OpenPriceFeedPayload[]> {
+  let promises = await Promise.allSettled(sources.map(async (sourceRaw) => {
+    let source = sourceRaw.includes('{') ? JSON.parse(sourceRaw) : sourceRaw;
```
Comment (coburncoburn) [inline poster/src/poster.ts:116]: nice this extra parsing is useful for wrong content-type on response

File: poster/src/poster.ts [hunk 4]
```diff
@@ -2,224 +2,173 @@ import { postWithRetries } from './post_with_retries';
...
+    .forEach(({source, symbol, price, prev}) => {
+      console.log(`Setting Price: source=${source}, symbol=${symbol}, price=${price}, prev_price=${prev}`);
+    });
+
+  return feedItems;
 }
 
 // Checks if new price is less than delta percent different form the old price
-function inDeltaRange(delta: number, price: number, prev_price: number) {
+// Note TODO: price here is uh... a number that needs to be scaled by 1e6?
```
Comment (antonina-cherednichenko) [inline poster/src/poster.ts:102]: Yes, the price from coinbase api needs to be scaled to 1e6, it's coming in this format:
"BTC": "9212.115", "ETH": "238.03", "XTZ": "3.04975" ...
and prevPrice is scaled to1e6 already.

Comment (hayesgm) [reply poster/src/poster.ts:102]: Yup, but I think the answer is: why is this code written that it accepts two numbers but assumes they are of different scales? I fixed this by descaling the price before this function is called.

File: poster/src/poster.ts [hunk 5]
```diff
@@ -2,224 +2,173 @@ import { postWithRetries } from './post_with_retries';
...
-  assets: string,
-  web3: Web3) {
-  const payloads = await fetchPayloads(sources.split(","));
-  const filteredPayloads = await filterPayloads(payloads, viewAddress, assets, delta, web3);
-  if (filteredPayloads.length != 0) {
+import { CoinbaseConfig, readCoinbasePayload } from './sources/coinbase';
+import { decodeMessage, encode, zip } from './util';
+
+const GAS_PRICE_API = 'https://api.compound.finance/api/gas_prices/get_gas_price';
+const DEFAULT_GAS_PRICE = 3_000_000_000; // use 3 gwei if api is unreachable for some reason
```
Comment (maxwolff) [inline poster/src/poster.ts:15]: maybe 30 gwei?

File: .circleci/config.yml [hunk 6]
```diff
@@ -51,7 +51,7 @@ jobs:
 
   test_poster:
     docker:
-      - image: circleci/node:11
+      - image: circleci/node:12
```
Comment (antonina-cherednichenko) [inline .circleci/config.yml:54]: node:13 here?

**Pull request merged**